### PR TITLE
For #15369: Add synced tabs metrics.

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by mcarare** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

_1) What questions will you answer with this data?_
How often users open a synced tab.

_2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?_
Measure whether this feature is used by users and how often.

_3) What alternative methods did you consider to answer these questions? Why were they not sufficient?_
This is the only way.

_4) Can current instrumentation answer these questions?_
No. We need a new specific event for a specific app feature.

_5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki._

_**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**_

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Open synced tab events</td>
    <td>Category 2</td>
    <td>https://github.com/mozilla-mobile/fenix/issues/15369</td>
  </tr>
</table>

_6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way._
This collection is documented in the `metrics.md` document document included in the project - https://github.com/mozilla-mobile/fenix/blob/master/docs/metrics.md

_7) How long will this data be collected?  Choose one of the following:_
Until 2021-04-01.

_8) What populations will you measure?_
All release channels and locales.

_9) If this data collection is default on, what is the opt-out mechanism for users?_
 Users can opt of of data collection by disabling Usage and technical data from Settings -> Data collection.

_10) Please provide a general description of how you will analyze this data._
 Glean / Amplitude

_11) Where do you intend to share the results of your analysis?_
 Only on Glean, Amplitude and with mobile teams.

_12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?_
No.

File: app/src/main/java/org/mozilla/fenix/tabtray/TabTrayController.kt [hunk 1]
```diff
@@ -205,6 +205,7 @@ class DefaultTabTrayController(
             newTab = true,
             from = BrowserDirection.FromTabTray
         )
+        metrics.track(Event.SyncedTabOpened)
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/tabtray/TabTrayController.kt:208]: The implementation of the Synced Tabs here is behind a secret settings flag and is also going away, so the metrics here would be skewed.

Maybe we wanted to add it to the three-dot menu?

File: app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt [hunk 2]
```diff
@@ -659,6 +659,12 @@ private val Event.wrapper: EventWrapper<*>?
             { ProgressiveWebApp.backgroundKeys.valueOf(it) }
         )
 
+        Event.SyncedTabOpened -> EventWrapper<NoExtraKeys>(
+            { Metrics.syncedTabsOpenedCount.add(1) }
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt:663]: > A counter that indicates how many synced tabs a user has opened

The description for this event reads like we want to count the number of times a synced tab _item_ is clicked and not the synced tabs UI is opened. Is that right?

If so, we probably want to add this counter to where the [tab was clicked](https://github.com/mozilla-mobile/fenix/blob/77f061c362dfb64de563eee60ec8c694c8cf290e/app/src/main/java/org/mozilla/fenix/sync/SyncedTabsLayout.kt#L120).

**Review by jonalmeida — CHANGES_REQUESTED** [review]
My comments went as individual items by mistake, sorry about the notifications!

**Comment by mcarare** [comment]
@jonalmeida Responding here, because I am not sure the right code was referenced in the comments.
I added the event tracking in `TabTrayController` `handleSyncedTabClicked` because this is triggered when clicking on a synced tab, it also contains the code to open that tab in the browser.
This is called by `ListenerDelegate`.`onTabClicked` but I did not want to trigger event tracking there because of the comment 

>         // Sanity-check: Remove this class when the feature flag is always enabled.
and I did not know how that would affect the code in `ListenerDelegate`.

**Comment by jonalmeida** [comment]
> I added the event tracking in `TabTrayController` `handleSyncedTabClicked` because this is triggered when clicking on a synced tab, it also contains the code to open that tab in the browser.

Ordinarily, this would be true. Although, Synced Tabs in the tabs tray is behind a flag in the "secret settings" so this metric would not be fired (unless a power user knows about this setting and enables it). The default location for the UI exists in the browser menu, so the data we collect from this metric would probably be skewed to those power users then. It might make more sense to track the metric in the [`BrowserToolbarMenuController`](https://github.com/mozilla-mobile/fenix/blob/57ebc0cb999180a2b2e100f1867f40470198be33/app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarMenuController.kt#L126) instead. What do you think?

Another part I would like to understand is the `synced_tabs_opened_count` event in `metrics.md`, if I understand correctly, that is tracked with:
```kotlin
Event.SyncedTabOpened -> EventWrapper<NoExtraKeys>(
    { Metrics.syncedTabsOpenedCount.add(1) }
)
```
This seems to coupling two events together: opening the Synced Tabs UI and clicking on synced tab items. If I understand that correctly, then similar to the above situation, we may be better tracking this in the `SyncedTabsLayout` which is used in the fragment for the Synced Tabs fragment.

When we move Synced Tabs to a new place in the tabs tray, we should definitely fix the metrics into the [re-written tabs tray](https://github.com/mozilla-mobile/fenix/issues/14969).

Apologises if I've caused any confusion. 🙂

File: app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt (hunk 2 above)
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt:663]: We spoke offline and clarified how metrics are recorded.

**Review by jonalmeida — APPROVED** [review]
We spoke offline to clarify the questions. Thanks!

**Comment by Mugurell** [comment]
Talked offline with Kate who confirmed with DS and Boek:
- The newly added metrics should log events, not counters anymore
- An an expiration date 6 months is preferred.
Maybe if there is already related telemetry for that part of the screen / feature we can reuse the existing expiration date so that all related telemetry probes will be renewed at the same time.

**Review by boek — APPROVED** [review]
# Data Review Form (to be filled by Data Stewards)

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?
Yes, metrics.yaml and metrics.md

2) Is there a control mechanism that allows the user to turn the data collection on and off?
Yes, Fenix data collection settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?
No, Fenix team will monitor

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?
Type 2

5) Is the data collection request for default-on or default-off?
Default on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?
No

7) Is the data collection covered by the existing Firefox privacy notice?
Yes

8) Does there need to be a check-in in the future to determine whether to renew the data?
Has expiry

9) Does the data collection use a third-party collection tool?
No

**Pull request merged**